ingestion_server/ingestion_server/cleanup.py

```diff
@@ -273,25 +284,68 @@ def _clean_data_worker(rows, temp_table, sources_config, all_fields: list[str]):
 
 
 def save_cleaned_data(result: dict) -> dict[str, int]:
-    log.info("Saving cleaned data...")
     start_time = time.perf_counter()
 
     cleanup_counts = {field: len(items) for field, items in result.items()}
     for field, cleaned_items in result.items():
         # Skip the tag field because the file is too large and fills up the disk
-        if field == "tag":
+        if field == "tag" or not cleaned_items:
             continue
-        if cleaned_items:
-            with open(f"{field}.tsv", "a") as f:
-                csv_writer = csv.writer(f, delimiter="\t")
-                csv_writer.writerows(cleaned_items)
+
+        with open(TMP_DIR.joinpath(f"{field}.tsv"), "a", encoding="utf-8") as f:
+            csv_writer = csv.writer(f, delimiter="\t")
+            csv_writer.writerows(cleaned_items)
 
     end_time = time.perf_counter()
     total_time = end_time - start_time
     log.info(f"Finished saving cleaned data in {total_time:.3f},\n{cleanup_counts}")
     return cleanup_counts
 
 
+def _get_s3_resource():
+    """
+    Locally, it connects to a MinIO instance through its endpoint and test credentials.
+    On live environments, the connection is allowed via IAM roles.
+    """
+    if config("ENVIRONMENT", default="local") == "local":
+        return boto3.resource(
+            "s3",
+            endpoint_url=config("AWS_S3_ENDPOINT", default="http://s3:5000"),
+            aws_access_key_id=config("AWS_ACCESS_KEY_ID", default="test_key"),
+            aws_secret_access_key=config(
+                "AWS_SECRET_ACCESS_KEY", default="test_secret"
+            ),
+        )
+
+    return boto3.resource("s3", region_name=config("AWS_REGION", default="us-east-1"))
```

Comment:
```suggestion
    if config("ENVIRONMENT", default="local") == "local":
        return boto3.resource(
            "s3",
            endpoint_url=config("AWS_S3_ENDPOINT", default="http://s3:5000"),
            aws_access_key_id=config("AWS_ACCESS_KEY_ID", default="test_key"),
            aws_secret_access_key=config(
                "AWS_SECRET_ACCESS_KEY", default="test_secret"
            ),
        )

    return boto3.resource(
        "s3",
        endpoint_url=config("AWS_S3_ENDPOINT", default=None),
        aws_access_key_id=config("AWS_ACCESS_KEY_ID", default=None),
        aws_secret_access_key=config("AWS_SECRET_ACCESS_KEY", default=None),
        region_name=config("AWS_REGION", default=None),
    )
```

All these kwargs are defined with None as the default in the implementation, so we can safely explicitly pass null (if these variables aren't defined in the environment), and they'll just be ignored.

https://boto3.amazonaws.com/v1/documentation/api/latest/_modules/boto3/session.html#Session.resource

That way the code doesn't have to worry about implementation details of local/live environments, and we get rid of this confusion altogether...

Reply:
That's a good observation! Thank you. I agree that we can simplify this a lot, though it'd require setting these environment variables locally for the ingestion server to work.

Reply:
Moved to #4529.